elf_reader.go

```diff
@@ -393,19 +395,51 @@ func (ec *elfCode) loadFunctions(section *elfSection) (map[string]asm.Instructio
 			// When splitting sections into subprograms, the targets of these calls
 			// are no longer in scope, so they must be resolved here.
 			if ins.IsFunctionReference() && ins.Constant != -1 {
-				tgt := jumpTarget(offset, ins)
+				tgt := jumpTarget(secOffset, ins)
 				sym := section.symbols[tgt].Name
 				if sym == "" {
-					return nil, fmt.Errorf("offset %d: no jump target found at offset %d", offset, tgt)
+					return nil, fmt.Errorf("offset %d: no jump target found at offset %d", secOffset, tgt)
 				}
 
-				ins.Reference = sym
+				ins.SetReference(sym)
 				ins.Constant = -1
 			}
 		}
 
+		// Only rebuild lineInfos at the start of a function
+		if funcOffset == 0 {
+			lineInfos = make(map[uint64]btf.LineInfo)
+			if ec.btf != nil {
+				for _, lineInfo := range ec.btf.GetFuncLineInfos(ins.Symbol()) {
+					lineInfos[uint64(lineInfo.InsnOff)] = lineInfo
+				}
+			}
+		}
+
+		if lineInfo, found := lineInfos[funcOffset]; found {
+			fileName, err := lineInfo.FileName(ec.btf)
+			if err != nil {
+				fileName = ""
+			}
+
+			line, err := lineInfo.Line(ec.btf)
+			if err != nil {
+				line = ""
+			}
+
+			// TODO we should use the *btf.LineInfo here directly, as soon as it implements fmt.Stringer
```

Comment:
I don't think this is necessary. We should avoid adding any  dependencies to the `asm` package. Just a string is fine here (see my thoughts around `SimpleContext` above).